PGWatch3 does not respect source DB version for metric queries

**Describe the bug**
PGWatch3 does not seem to use PG13 queries and uses PG15 or PG16 queries for PG14 source.

**To Reproduce**
Steps to reproduce the behavior:
1. Add PG14 Source
2. Collect pg_stat_statements and table_stats
3. Check error log

**Expected behavior**
I expect no errors in log.

**Versions:**
pgwatch3:
  Version:      master
  DB Schema:    00179
  Git Commit:   742775e04f626792bf6906065f6c58bf68bb2e4a
  Built:        2024-07-23T17:27:23+02:00
postgres 
```
SELECT dbname, data->>'major_version' as major_version, data->>'server_version' as server_version, data->>'server_version_num' as server_version_num FROM settings LIMIT 1;
   dbname   | major_version | server_version | server_version_num
------------+---------------+----------------+--------------------
 Lab_zabbix | 14            | 14.12          | 140012
```

**Additional context**
Log entries since it uses PG 15 queries.
```
2024-07-24 09:56:42.933 [ERROR] [source:Lab_zabbix] [metric:stat_statements] [error:ERROR: column "temp_blk_read_time" does not exist (SQLSTATE 42703)] failed to fetch metric data
2024-07-24 09:56:42.962 [ERROR] [source:Lab_zabbix] [metric:table_stats] [error:ERROR: column "last_seq_scan" does not exist (SQLSTATE 42703)] failed to fetch metric data
```

--- a/src/reaper/cache.go
+++ b/src/reaper/cache.go
@@ -14,9 +14,9 @@ import (
 
 var monitoredDbCache map[string]*sources.MonitoredDatabase
 var monitoredDbCacheLock sync.RWMutex
-var dbPgVersionMap = make(map[string]DBVersionMapEntry)
-var dbPgVersionMapLock = sync.RWMutex{}
-var dbGetPgVersionMapLock = make(map[string]*sync.RWMutex) // synchronize initial PG version detection to 1 instance for each defined host
+var MonitoredDatabasesSettings = make(map[string]MonitoredDatabaseSettings)
+var MonitoredDatabasesSettingsLock = sync.RWMutex{}
+var MonitoredDatabasesSettingsGetLock = make(map[string]*sync.RWMutex) // synchronize initial PG version detection to 1 instance for each defined host
 
 var unreachableDBsLock sync.RWMutex
 var unreachableDB = make(map[string]time.Time)
@@ -36,11 +36,11 @@ var hostMetricIntervalMap = make(map[string]float64) // [db1_metric] = 30
 var lastSQLFetchError sync.Map
 
 func InitPGVersionInfoFetchingLockIfNil(md *sources.MonitoredDatabase) {
-	dbPgVersionMapLock.Lock()
-	if _, ok := dbGetPgVersionMapLock[md.DBUniqueName]; !ok {
-		dbGetPgVersionMapLock[md.DBUniqueName] = &sync.RWMutex{}
+	MonitoredDatabasesSettingsLock.Lock()
+	if _, ok := MonitoredDatabasesSettingsGetLock[md.DBUniqueName]; !ok {
+		MonitoredDatabasesSettingsGetLock[md.DBUniqueName] = &sync.RWMutex{}
 	}
-	dbPgVersionMapLock.Unlock()
+	MonitoredDatabasesSettingsLock.Unlock()
 }
 
 func UpdateMonitoredDBCache(data sources.MonitoredDatabases) {
@@ -64,7 +64,7 @@ func GetMonitoredDatabaseByUniqueName(name string) (*sources.MonitoredDatabase, 
 }
 
 // assumes upwards compatibility for versions
-func GetMetricVersionProperties(metric string, _ DBVersionMapEntry, metricDefMap *metrics.Metrics) (metrics.Metric, error) {
+func GetMetricVersionProperties(metric string, _ MonitoredDatabaseSettings, metricDefMap *metrics.Metrics) (metrics.Metric, error) {
 	mdm := new(metrics.Metrics)
 	if metricDefMap != nil {
 		mdm = metricDefMap
@@ -106,7 +106,7 @@ func SyncMetricDefs(ctx context.Context, r metrics.Reader) {
 	}
 }
 
-func GetFromInstanceCacheIfNotOlderThanSeconds(msg MetricFetchMessage, maxAgeSeconds int64) metrics.Measurements {
+func GetFromInstanceCacheIfNotOlderThanSeconds(msg MetricFetchConfig, maxAgeSeconds int64) metrics.Measurements {
 	var clonedData metrics.Measurements
 	instanceMetricCacheTimestampLock.RLock()
 	instanceMetricTS, ok := instanceMetricCacheTimestamp[msg.DBUniqueNameOrig+msg.MetricName]

--- a/src/reaper/database.go
+++ b/src/reaper/database.go
@@ -99,7 +99,7 @@ func DBGetSizeMB(ctx context.Context, dbUnique string) (int64, error) {
 	lastDBSizeCheckLock.RUnlock()
 
 	if !ok || lastDBSizeCheckTime.Add(dbSizeCachingInterval).Before(time.Now()) {
-		ver, err := DBGetPGVersion(ctx, dbUnique, sources.SourcePostgres, false, "")
+		ver, err := GetMonitoredDatabaseSettings(ctx, dbUnique, sources.SourcePostgres, false)
 		if err != nil || (ver.ExecEnv != execEnvAzureSingle) || (ver.ExecEnv == execEnvAzureSingle && ver.ApproxDBSizeB < 1e12) {
 			log.GetLogger(ctx).Debugf("[%s] determining DB size ...", dbUnique)
 
@@ -128,21 +128,17 @@ func DBGetSizeMB(ctx context.Context, dbUnique string) (int64, error) {
 	return lastDBSize, nil
 }
 
-func TryDiscoverExecutionEnv(ctx context.Context, dbUnique string) string {
-	sqlPGExecEnv := `select /* pgwatch3_generated */
+func TryDiscoverExecutionEnv(ctx context.Context, dbUnique string) (execEnv string) {
+	sql := `select /* pgwatch3_generated */
 	case
 	  when exists (select * from pg_settings where name = 'pg_qs.host_database' and setting = 'azure_sys') and version() ~* 'compiled by Visual C' then 'AZURE_SINGLE'
 	  when exists (select * from pg_settings where name = 'pg_qs.host_database' and setting = 'azure_sys') and version() ~* 'compiled by gcc' then 'AZURE_FLEXIBLE'
 	  when exists (select * from pg_settings where name = 'cloudsql.supported_extensions') then 'GOOGLE'
 	else
 	  'UNKNOWN'
-	end as exec_env;
-  `
-	data, err := DBExecReadByDbUniqueName(ctx, dbUnique, sqlPGExecEnv)
-	if err != nil {
-		return ""
-	}
-	return data[0]["exec_env"].(string)
+	end as exec_env`
+	_ = GetConnByUniqueName(dbUnique).QueryRow(ctx, sql).Scan(&execEnv)
+	return
 }
 
 func GetDBTotalApproxSize(ctx context.Context, dbUnique string) (int64, error) {
@@ -176,162 +172,130 @@ func VersionToInt(version string) (v int) {
 	return
 }
 
-const MinExtensionInfoAvailable = 9_01_00
-
 var rBouncerAndPgpoolVerMatch = regexp.MustCompile(`\d+\.+\d+`) // extract $major.minor from "4.1.2 (karasukiboshi)" or "PgBouncer 1.12.0"
 
-func DBGetPGVersion(ctx context.Context, dbUnique string, srcType sources.Kind, noCache bool, SysID string) (DBVersionMapEntry, error) {
-	var ver DBVersionMapEntry
-	var verNew DBVersionMapEntry
+func GetMonitoredDatabaseSettings(ctx context.Context, dbUnique string, srcType sources.Kind, noCache bool) (MonitoredDatabaseSettings, error) {
+	var dbSettings MonitoredDatabaseSettings
+	var dbNewSettings MonitoredDatabaseSettings
 	var ok bool
-	sql := `
-		select /* pgwatch3_generated */ (regexp_matches(
-			regexp_replace(current_setting('server_version'), '(beta|devel).*', '', 'g'),
-			E'\\d+\\.?\\d+?')
-			)[1]::text as ver, pg_is_in_recovery(), current_database()::text;
-	`
-	sqlSysid := `select /* pgwatch3_generated */ system_identifier::text from pg_control_system();`
-	sqlSu := `select /* pgwatch3_generated */ rolsuper
-			   from pg_roles r where rolname = session_user;`
+
+	l := log.GetLogger(ctx).WithField("source", dbUnique).WithField("kind", srcType)
+
 	sqlExtensions := `select /* pgwatch3_generated */ extname::text, (regexp_matches(extversion, $$\d+\.?\d+?$$))[1]::text as extversion from pg_extension order by 1;`
-	pgpoolVersion := `SHOW POOL_VERSION` // supported from pgpool2 v3.0
 
-	dbPgVersionMapLock.Lock()
-	getVerLock, ok := dbGetPgVersionMapLock[dbUnique]
+	MonitoredDatabasesSettingsLock.Lock()
+	getVerLock, ok := MonitoredDatabasesSettingsGetLock[dbUnique]
 	if !ok {
-		dbGetPgVersionMapLock[dbUnique] = &sync.RWMutex{}
-		getVerLock = dbGetPgVersionMapLock[dbUnique]
+		MonitoredDatabasesSettingsGetLock[dbUnique] = &sync.RWMutex{}
+		getVerLock = MonitoredDatabasesSettingsGetLock[dbUnique]
 	}
-	ver, ok = dbPgVersionMap[dbUnique]
-	dbPgVersionMapLock.Unlock()
+	dbSettings, ok = MonitoredDatabasesSettings[dbUnique]
+	MonitoredDatabasesSettingsLock.Unlock()
 
-	if !noCache && ok && ver.LastCheckedOn.After(time.Now().Add(time.Minute*-2)) { // use cached version for 2 min
+	if !noCache && ok && dbSettings.LastCheckedOn.After(time.Now().Add(time.Minute*-2)) { // use cached version for 2 min
 		//log.Debugf("using cached postgres version %s for %s", ver.Version.String(), dbUnique)
-		return ver, nil
+		return dbSettings, nil
 	}
 	getVerLock.Lock() // limit to 1 concurrent version info fetch per DB
 	defer getVerLock.Unlock()
-	log.GetLogger(ctx).WithField("source", dbUnique).
-		WithField("type", srcType).Debug("determining DB version and recovery status...")
+	l.Debug("determining DB version and recovery status...")
 
-	if verNew.Extensions == nil {
-		verNew.Extensions = make(map[string]int)
+	if dbNewSettings.Extensions == nil {
+		dbNewSettings.Extensions = make(map[string]int)
 	}
 
-	if srcType == sources.SourcePgBouncer {
-		data, err := DBExecReadByDbUniqueName(ctx, dbUnique, "show version")
-		if err != nil {
-			return verNew, err
+	switch srcType {
+	case sources.SourcePgBouncer:
+		if err := GetConnByUniqueName(dbUnique).QueryRow(ctx, "SHOW VERSION").Scan(&dbNewSettings.VersionStr); err != nil {
+			return dbNewSettings, err
 		}
-		if len(data) == 0 {
-			// surprisingly pgbouncer 'show version' outputs in pre v1.12 is emitted as 'NOTICE' which cannot be accessed from Go lib/pg
-			verNew.Version = 0
-			verNew.VersionStr = "0"
-		} else {
-			matches := rBouncerAndPgpoolVerMatch.FindStringSubmatch(data[0]["version"].(string))
-			if len(matches) != 1 {
-				log.GetLogger(ctx).Errorf("[%s] Unexpected PgBouncer version input: %s", dbUnique, data[0]["version"].(string))
-				return ver, fmt.Errorf("Unexpected PgBouncer version input: %s", data[0]["version"].(string))
-			}
-			verNew.VersionStr = matches[0]
-			verNew.Version = VersionToInt(matches[0])
+		matches := rBouncerAndPgpoolVerMatch.FindStringSubmatch(dbNewSettings.VersionStr)
+		if len(matches) != 1 {
+			return dbSettings, fmt.Errorf("Unexpected PgBouncer version input: %s", dbNewSettings.VersionStr)
 		}
-	} else if srcType == sources.SourcePgPool {
-		data, err := DBExecReadByDbUniqueName(ctx, dbUnique, pgpoolVersion)
-		if err != nil {
-			return verNew, err
+		dbNewSettings.Version = VersionToInt(matches[0])
+	case sources.SourcePgPool:
+		if err := GetConnByUniqueName(dbUnique).QueryRow(ctx, "SHOW POOL_VERSION").Scan(&dbNewSettings.VersionStr); err != nil {
+			return dbNewSettings, err
 		}
-		if len(data) == 0 {
-			verNew.Version = 3_00_00
-			verNew.VersionStr = "3.0"
-		} else {
-			matches := rBouncerAndPgpoolVerMatch.FindStringSubmatch(string(data[0]["pool_version"].([]byte)))
-			if len(matches) != 1 {
-				log.GetLogger(ctx).Errorf("[%s] Unexpected PgPool version input: %s", dbUnique, data[0]["pool_version"].([]byte))
-				return ver, fmt.Errorf("Unexpected PgPool version input: %s", data[0]["pool_version"].([]byte))
-			}
-			verNew.VersionStr = matches[0]
-			verNew.Version = VersionToInt(matches[0])
+
+		matches := rBouncerAndPgpoolVerMatch.FindStringSubmatch(dbNewSettings.VersionStr)
+		if len(matches) != 1 {
+			return dbSettings, fmt.Errorf("Unexpected PgPool version input: %s", dbNewSettings.VersionStr)
 		}
-	} else {
-		data, err := DBExecReadByDbUniqueName(ctx, dbUnique, sql)
+		dbNewSettings.Version = VersionToInt(matches[0])
+	default:
+		sql := `select /* pgwatch3_generated */ 
+	current_setting('server_version_num')::int / 1_00_00 as ver, 
+	version(), 
+	pg_is_in_recovery(), 
+	current_database()::TEXT,
+	system_identifier
+FROM
+	pg_control_system()`
+
+		err := GetConnByUniqueName(dbUnique).QueryRow(ctx, sql).
+			Scan(&dbNewSettings.Version, &dbNewSettings.VersionStr,
+				&dbNewSettings.IsInRecovery, &dbNewSettings.RealDbname,
+				&dbNewSettings.SystemIdentifier)
 		if err != nil {
 			if noCache {
-				return ver, err
+				return dbSettings, err
 			}
-			log.GetLogger(ctx).Infof("[%s] DBGetPGVersion failed, using old cached value. err: %v", dbUnique, err)
-			return ver, nil
-
-		}
-		verNew.Version = VersionToInt(data[0]["ver"].(string))
-		verNew.VersionStr = data[0]["ver"].(string)
-		verNew.IsInRecovery = data[0]["pg_is_in_recovery"].(bool)
-		verNew.RealDbname = data[0]["current_database"].(string)
-
-		if verNew.Version > 100000 && SysID > "" {
-			log.GetLogger(ctx).Debugf("[%s] determining system identifier version (pg ver: %v)", dbUnique, verNew.VersionStr)
-			data, err := DBExecReadByDbUniqueName(ctx, dbUnique, sqlSysid)
-			if err == nil && len(data) > 0 {
-				verNew.SystemIdentifier = data[0]["system_identifier"].(string)
-			}
+			l.Error("DBGetPGVersion failed, using old cached value: ", err)
+			return dbSettings, nil
 		}
 
-		if ver.ExecEnv != "" {
-			verNew.ExecEnv = ver.ExecEnv // carry over as not likely to change ever
+		if dbSettings.ExecEnv != "" {
+			dbNewSettings.ExecEnv = dbSettings.ExecEnv // carry over as not likely to change ever
 		} else {
-			log.GetLogger(ctx).Debugf("[%s] determining the execution env...", dbUnique)
-			execEnv := TryDiscoverExecutionEnv(ctx, dbUnique)
-			if execEnv != "" {
-				log.GetLogger(ctx).Debugf("[%s] running on execution env: %s", dbUnique, execEnv)
-				verNew.ExecEnv = execEnv
-			}
+			l.Debugf("determining the execution env...")
+			dbNewSettings.ExecEnv = TryDiscoverExecutionEnv(ctx, dbUnique)
 		}
 
 		// to work around poor Azure Single Server FS functions performance for some metrics + the --min-db-size-mb filter
-		if verNew.ExecEnv == execEnvAzureSingle {
-			approxSize, err := GetDBTotalApproxSize(ctx, dbUnique)
-			if err == nil {
-				verNew.ApproxDBSizeB = approxSize
+		if dbNewSettings.ExecEnv == execEnvAzureSingle {
+			if approxSize, err := GetDBTotalApproxSize(ctx, dbUnique); err == nil {
+				dbNewSettings.ApproxDBSizeB = approxSize
 			} else {
-				verNew.ApproxDBSizeB = ver.ApproxDBSizeB
+				dbNewSettings.ApproxDBSizeB = dbSettings.ApproxDBSizeB
 			}
 		}
 
-		log.GetLogger(ctx).Debugf("[%s] determining if monitoring user is a superuser...", dbUnique)
-		data, err = DBExecReadByDbUniqueName(ctx, dbUnique, sqlSu)
-		if err == nil {
-			verNew.IsSuperuser = data[0]["rolsuper"].(bool)
-		}
-		log.GetLogger(ctx).Debugf("[%s] superuser=%v", dbUnique, verNew.IsSuperuser)
+		l.Debugf("[%s] determining if monitoring user is a superuser...", dbUnique)
+		sqlSu := `select /* pgwatch3_generated */ rolsuper from pg_roles r where rolname = session_user`
 
-		if verNew.Version >= MinExtensionInfoAvailable {
-			//log.Debugf("[%s] determining installed extensions info...", dbUnique)
-			data, err = DBExecReadByDbUniqueName(ctx, dbUnique, sqlExtensions)
-			if err != nil {
-				log.GetLogger(ctx).Errorf("[%s] failed to determine installed extensions info: %v", dbUnique, err)
-			} else {
-				for _, dr := range data {
-					extver := VersionToInt(dr["extversion"].(string))
-					if extver == 0 {
-						log.GetLogger(ctx).Error("[%s] failed to determine extension version info for extension %s: %v", dbUnique, dr["extname"])
-						continue
-					}
-					verNew.Extensions[dr["extname"].(string)] = extver
+		if err = GetConnByUniqueName(dbUnique).QueryRow(ctx, sqlSu).Scan(&dbNewSettings.IsSuperuser); err != nil {
+			l.Errorf("[%s] failed to determine if monitoring user is a superuser: %v", dbUnique, err)
+		}
+
+		l.Debugf("[%s] determining installed extensions info...", dbUnique)
+		data, err := DBExecReadByDbUniqueName(ctx, dbUnique, sqlExtensions)
+		if err != nil {
+			l.Errorf("[%s] failed to determine installed extensions info: %v", dbUnique, err)
+		} else {
+			for _, dr := range data {
+				extver := VersionToInt(dr["extversion"].(string))
+				if extver == 0 {
+					l.Error("failed to determine extension version info for extension: ", dr["extname"])
+					continue
 				}
-				log.GetLogger(ctx).Debugf("[%s] installed extensions: %+v", dbUnique, verNew.Extensions)
+				dbNewSettings.Extensions[dr["extname"].(string)] = extver
 			}
+			l.Debugf("[%s] installed extensions: %+v", dbUnique, dbNewSettings.Extensions)
 		}
+
 	}
 
-	verNew.LastCheckedOn = time.Now()
-	dbPgVersionMapLock.Lock()
-	dbPgVersionMap[dbUnique] = verNew
-	dbPgVersionMapLock.Unlock()
+	dbNewSettings.LastCheckedOn = time.Now()
+	MonitoredDatabasesSettingsLock.Lock()
+	MonitoredDatabasesSettings[dbUnique] = dbNewSettings
+	MonitoredDatabasesSettingsLock.Unlock()
 
-	return verNew, nil
+	return dbNewSettings, nil
 }
 
-func DetectSprocChanges(ctx context.Context, dbUnique string, vme DBVersionMapEntry, storageCh chan<- []metrics.MeasurementMessage, hostState map[string]map[string]string) ChangeDetectionResults {
+func DetectSprocChanges(ctx context.Context, dbUnique string, vme MonitoredDatabaseSettings, storageCh chan<- []metrics.MeasurementMessage, hostState map[string]map[string]string) ChangeDetectionResults {
 	detectedChanges := make(metrics.Measurements, 0)
 	var firstRun bool
 	var changeCounts ChangeDetectionResults
@@ -415,7 +379,7 @@ func DetectSprocChanges(ctx context.Context, dbUnique string, vme DBVersionMapEn
 	return changeCounts
 }
 
-func DetectTableChanges(ctx context.Context, dbUnique string, vme DBVersionMapEntry, storageCh chan<- []metrics.MeasurementMessage, hostState map[string]map[string]string) ChangeDetectionResults {
+func DetectTableChanges(ctx context.Context, dbUnique string, vme MonitoredDatabaseSettings, storageCh chan<- []metrics.MeasurementMessage, hostState map[string]map[string]string) ChangeDetectionResults {
 	detectedChanges := make(metrics.Measurements, 0)
 	var firstRun bool
 	var changeCounts ChangeDetectionResults
@@ -499,7 +463,7 @@ func DetectTableChanges(ctx context.Context, dbUnique string, vme DBVersionMapEn
 	return changeCounts
 }
 
-func DetectIndexChanges(ctx context.Context, dbUnique string, vme DBVersionMapEntry, storageCh chan<- []metrics.MeasurementMessage, hostState map[string]map[string]string) ChangeDetectionResults {
+func DetectIndexChanges(ctx context.Context, dbUnique string, vme MonitoredDatabaseSettings, storageCh chan<- []metrics.MeasurementMessage, hostState map[string]map[string]string) ChangeDetectionResults {
 	detectedChanges := make(metrics.Measurements, 0)
 	var firstRun bool
 	var changeCounts ChangeDetectionResults
@@ -581,7 +545,7 @@ func DetectIndexChanges(ctx context.Context, dbUnique string, vme DBVersionMapEn
 	return changeCounts
 }
 
-func DetectPrivilegeChanges(ctx context.Context, dbUnique string, vme DBVersionMapEntry, storageCh chan<- []metrics.MeasurementMessage, hostState map[string]map[string]string) ChangeDetectionResults {
+func DetectPrivilegeChanges(ctx context.Context, dbUnique string, vme MonitoredDatabaseSettings, storageCh chan<- []metrics.MeasurementMessage, hostState map[string]map[string]string) ChangeDetectionResults {
 	detectedChanges := make(metrics.Measurements, 0)
 	var firstRun bool
 	var changeCounts ChangeDetectionResults
@@ -663,7 +627,7 @@ func DetectPrivilegeChanges(ctx context.Context, dbUnique string, vme DBVersionM
 	return changeCounts
 }
 
-func DetectConfigurationChanges(ctx context.Context, dbUnique string, vme DBVersionMapEntry, storageCh chan<- []metrics.MeasurementMessage, hostState map[string]map[string]string) ChangeDetectionResults {
+func DetectConfigurationChanges(ctx context.Context, dbUnique string, vme MonitoredDatabaseSettings, storageCh chan<- []metrics.MeasurementMessage, hostState map[string]map[string]string) ChangeDetectionResults {
 	detectedChanges := make(metrics.Measurements, 0)
 	var firstRun bool
 	var changeCounts ChangeDetectionResults
@@ -727,7 +691,7 @@ func DetectConfigurationChanges(ctx context.Context, dbUnique string, vme DBVers
 	return changeCounts
 }
 
-func CheckForPGObjectChangesAndStore(ctx context.Context, dbUnique string, vme DBVersionMapEntry, storageCh chan<- []metrics.MeasurementMessage, hostState map[string]map[string]string) {
+func CheckForPGObjectChangesAndStore(ctx context.Context, dbUnique string, vme MonitoredDatabaseSettings, storageCh chan<- []metrics.MeasurementMessage, hostState map[string]map[string]string) {
 	sprocCounts := DetectSprocChanges(ctx, dbUnique, vme, storageCh, hostState) // TODO some of Detect*() code could be unified...
 	tableCounts := DetectTableChanges(ctx, dbUnique, vme, storageCh, hostState)
 	indexCounts := DetectIndexChanges(ctx, dbUnique, vme, storageCh, hostState)
@@ -772,7 +736,7 @@ func CheckForPGObjectChangesAndStore(ctx context.Context, dbUnique string, vme D
 }
 
 // some extra work needed as pgpool SHOW commands don't specify the return data types for some reason
-func FetchMetricsPgpool(ctx context.Context, msg MetricFetchMessage, vme DBVersionMapEntry, mvp metrics.Metric) (metrics.Measurements, error) {
+func FetchMetricsPgpool(ctx context.Context, msg MetricFetchConfig, vme MonitoredDatabaseSettings, mvp metrics.Metric) (metrics.Measurements, error) {
 	var retData = make(metrics.Measurements, 0)
 	epochNs := time.Now().UnixNano()
 

--- a/src/reaper/file.go
+++ b/src/reaper/file.go
@@ -35,7 +35,7 @@ func IsDirectlyFetchableMetric(metric string) bool {
 	return ok
 }
 
-func FetchStatsDirectlyFromOS(ctx context.Context, msg MetricFetchMessage, vme DBVersionMapEntry, mvp metrics.Metric) ([]metrics.MeasurementMessage, error) {
+func FetchStatsDirectlyFromOS(ctx context.Context, msg MetricFetchConfig, vme MonitoredDatabaseSettings, mvp metrics.Metric) ([]metrics.MeasurementMessage, error) {
 	var data []map[string]any
 	var err error
 
@@ -62,7 +62,7 @@ func FetchStatsDirectlyFromOS(ctx context.Context, msg MetricFetchMessage, vme D
 }
 
 // data + custom tags + counters
-func DatarowsToMetricstoreMessage(data metrics.Measurements, msg MetricFetchMessage, vme DBVersionMapEntry, mvp metrics.Metric) (metrics.MeasurementMessage, error) {
+func DatarowsToMetricstoreMessage(data metrics.Measurements, msg MetricFetchConfig, vme MonitoredDatabaseSettings, mvp metrics.Metric) (metrics.MeasurementMessage, error) {
 	md, err := GetMonitoredDatabaseByUniqueName(msg.DBUniqueName)
 	if err != nil {
 		return metrics.MeasurementMessage{}, err

--- a/src/reaper/reaper.go
+++ b/src/reaper/reaper.go
@@ -120,9 +120,9 @@ func (r *Reaper) Reap(mainContext context.Context) (err error) {
 
 			InitPGVersionInfoFetchingLockIfNil(monitoredDB)
 
-			var ver DBVersionMapEntry
+			var ver MonitoredDatabaseSettings
 
-			ver, err = DBGetPGVersion(mainContext, dbUnique, srcType, true, opts.Measurements.SystemIdentifierField)
+			ver, err = GetMonitoredDatabaseSettings(mainContext, dbUnique, srcType, true)
 			if err != nil {
 				logger.Errorf("could not start metric gathering for DB \"%s\" due to connection problem: %s", dbUnique, err)
 				continue
@@ -180,7 +180,7 @@ func (r *Reaper) Reap(mainContext context.Context) (err error) {
 						SetUndersizedDBState(dbUnique, false)
 					}
 				}
-				ver, err := DBGetPGVersion(mainContext, dbUnique, monitoredDB.Kind, false, opts.Measurements.SystemIdentifierField)
+				ver, err := GetMonitoredDatabaseSettings(mainContext, dbUnique, monitoredDB.Kind, false)
 				if err == nil { // ok to ignore error, re-tried on next loop
 					lastKnownStatusInRecovery := hostLastKnownStatusInRecovery[dbUnique]
 					if ver.IsInRecovery && monitoredDB.OnlyIfMaster {
@@ -243,7 +243,7 @@ func (r *Reaper) Reap(mainContext context.Context) (err error) {
 
 						metricNameForStorage := metricName
 						if _, isSpecialMetric := specialMetrics[metricName]; !isSpecialMetric {
-							vme, err := DBGetPGVersion(mainContext, dbUnique, srcType, false, opts.Measurements.SystemIdentifierField)
+							vme, err := GetMonitoredDatabaseSettings(mainContext, dbUnique, srcType, false)
 							if err != nil {
 								logger.Warning("Failed to determine possible re-routing name, Grafana dashboards with re-routed metrics might not show all hosts")
 							} else {
@@ -319,9 +319,9 @@ func (r *Reaper) Reap(mainContext context.Context) (err error) {
 			}
 
 			if !(wholeDbShutDownDueToRoleChange || dbRemovedFromConfig) { // maybe some single metric was disabled
-				dbPgVersionMapLock.RLock()
-				verInfo, ok := dbPgVersionMap[db]
-				dbPgVersionMapLock.RUnlock()
+				MonitoredDatabasesSettingsLock.RLock()
+				verInfo, ok := MonitoredDatabasesSettings[db]
+				MonitoredDatabasesSettingsLock.RUnlock()
 				if !ok {
 					logger.Warningf("Could not find PG version info for DB %s, skipping shutdown check of metric worker process for %s", db, metric)
 					continue
@@ -385,7 +385,7 @@ func (r *Reaper) reapMetricMeasurementsFromSource(ctx context.Context,
 	hostState := make(map[string]map[string]string)
 	var lastUptimeS int64 = -1 // used for "server restarted" event detection
 	var lastErrorNotificationTime time.Time
-	var vme DBVersionMapEntry
+	var vme MonitoredDatabaseSettings
 	var mvp metrics.Metric
 	var err error
 	failedFetches := 0
@@ -397,9 +397,9 @@ func (r *Reaper) reapMetricMeasurementsFromSource(ctx context.Context,
 		if err != nil {
 			return
 		}
-		dbPgVersionMapLock.RLock()
-		realDbname := dbPgVersionMap[dbUniqueName].RealDbname // to manage 2 sets of event counts - monitored DB + global
-		dbPgVersionMapLock.RUnlock()
+		MonitoredDatabasesSettingsLock.RLock()
+		realDbname := MonitoredDatabasesSettings[dbUniqueName].RealDbname // to manage 2 sets of event counts - monitored DB + global
+		MonitoredDatabasesSettingsLock.RUnlock()
 		conn := mdb.Conn
 		metrics.ParseLogs(ctx, conn, mdb, realDbname, metricName, configMap, r.measurementCh) // no return
 		return
@@ -408,7 +408,7 @@ func (r *Reaper) reapMetricMeasurementsFromSource(ctx context.Context,
 	for {
 		interval := configMap[metricName]
 		if lastDBVersionFetchTime.Add(time.Minute * time.Duration(5)).Before(time.Now()) {
-			vme, err = DBGetPGVersion(ctx, dbUniqueName, srcType, false, r.opts.Measurements.SystemIdentifierField) // in case of errors just ignore metric "disabled" time ranges
+			vme, err = GetMonitoredDatabaseSettings(ctx, dbUniqueName, srcType, false) // in case of errors just ignore metric "disabled" time ranges
 			if err != nil {
 				lastDBVersionFetchTime = time.Now()
 			}
@@ -422,7 +422,7 @@ func (r *Reaper) reapMetricMeasurementsFromSource(ctx context.Context,
 
 		var metricStoreMessages []metrics.MeasurementMessage
 		var err error
-		mfm := MetricFetchMessage{
+		mfm := MetricFetchConfig{
 			DBUniqueName:        dbUniqueName,
 			DBUniqueNameOrig:    dbUniqueNameOrig,
 			MetricName:          metricName,
@@ -537,7 +537,7 @@ func SyncMonitoredDBsToDatastore(ctx context.Context, monitoredDbs []*sources.Mo
 	}
 }
 
-func AddDbnameSysinfoIfNotExistsToQueryResultData(data metrics.Measurements, ver DBVersionMapEntry, opts *config.Options) metrics.Measurements {
+func AddDbnameSysinfoIfNotExistsToQueryResultData(data metrics.Measurements, ver MonitoredDatabaseSettings, opts *config.Options) metrics.Measurements {
 	enrichedData := make(metrics.Measurements, 0)
 	for _, dr := range data {
 		if opts.Measurements.RealDbnameField > "" && ver.RealDbname > "" {
@@ -563,14 +563,14 @@ var instanceMetricCacheLock = sync.RWMutex{}
 var instanceMetricCacheTimestamp = make(map[string]time.Time) // [dbUnique+metric]last_fetch_time
 var instanceMetricCacheTimestampLock = sync.RWMutex{}
 
-func IsCacheableMetric(msg MetricFetchMessage, mvp metrics.Metric) bool {
+func IsCacheableMetric(msg MetricFetchConfig, mvp metrics.Metric) bool {
 	if !(msg.Source == sources.SourcePostgresContinuous || msg.Source == sources.SourcePatroniContinuous) {
 		return false
 	}
 	return mvp.IsInstanceLevel
 }
 
-func PutToInstanceCache(msg MetricFetchMessage, data metrics.Measurements) {
+func PutToInstanceCache(msg MetricFetchConfig, data metrics.Measurements) {
 	if len(data) == 0 {
 		return
 	}
@@ -599,14 +599,14 @@ func deepCopyMetricData(data metrics.Measurements) metrics.Measurements {
 }
 
 func FetchMetrics(ctx context.Context,
-	msg MetricFetchMessage,
+	msg MetricFetchConfig,
 	hostState map[string]map[string]string,
 	storageCh chan<- []metrics.MeasurementMessage,
 	context string,
 	opts *config.Options) ([]metrics.MeasurementMessage, error) {
 
-	var vme DBVersionMapEntry
-	var dbpgVersion int
+	var dbSettings MonitoredDatabaseSettings
+	var dbVersion int
 	var err error
 	var sql string
 	var data, cachedData metrics.Measurements
@@ -619,33 +619,33 @@ func FetchMetrics(ctx context.Context,
 		return nil, err
 	}
 
-	vme, err = DBGetPGVersion(ctx, msg.DBUniqueName, msg.Source, false, opts.Measurements.SystemIdentifierField)
+	dbSettings, err = GetMonitoredDatabaseSettings(ctx, msg.DBUniqueName, msg.Source, false)
 	if err != nil {
 		log.GetLogger(ctx).Error("failed to fetch pg version for ", msg.DBUniqueName, msg.MetricName, err)
 		return nil, err
 	}
 	if msg.MetricName == specialMetricDbSize || msg.MetricName == specialMetricTableStats {
-		if vme.ExecEnv == execEnvAzureSingle && vme.ApproxDBSizeB > 1e12 { // 1TB
+		if dbSettings.ExecEnv == execEnvAzureSingle && dbSettings.ApproxDBSizeB > 1e12 { // 1TB
 			subsMetricName := msg.MetricName + "_approx"
-			mvpApprox, err := GetMetricVersionProperties(subsMetricName, vme, nil)
+			mvpApprox, err := GetMetricVersionProperties(subsMetricName, dbSettings, nil)
 			if err == nil && mvpApprox.StorageName == msg.MetricName {
 				log.GetLogger(ctx).Infof("[%s:%s] Transparently swapping metric to %s due to hard-coded rules...", msg.DBUniqueName, msg.MetricName, subsMetricName)
 				msg.MetricName = subsMetricName
 			}
 		}
 	}
-	dbpgVersion = vme.Version
+	dbVersion = dbSettings.Version
 
 	if msg.Source == sources.SourcePgBouncer {
-		dbpgVersion = 0 // version is 0.0 for all pgbouncer sql per convention
+		dbVersion = 0 // version is 0.0 for all pgbouncer sql per convention
 	}
 
-	mvp, err := GetMetricVersionProperties(msg.MetricName, vme, nil)
+	mvp, err := GetMetricVersionProperties(msg.MetricName, dbSettings, nil)
 	if err != nil && msg.MetricName != recoMetricName {
-		epoch, ok := lastSQLFetchError.Load(msg.MetricName + dbMetricJoinStr + fmt.Sprintf("%v", dbpgVersion))
+		epoch, ok := lastSQLFetchError.Load(msg.MetricName + dbMetricJoinStr + fmt.Sprintf("%v", dbVersion))
 		if !ok || ((time.Now().Unix() - epoch.(int64)) > 3600) { // complain only 1x per hour
-			log.GetLogger(ctx).Infof("Failed to get SQL for metric '%s', version '%s': %v", msg.MetricName, vme.VersionStr, err)
-			lastSQLFetchError.Store(msg.MetricName+dbMetricJoinStr+fmt.Sprintf("%v", dbpgVersion), time.Now().Unix())
+			log.GetLogger(ctx).Infof("Failed to get SQL for metric '%s', version '%s': %v", msg.MetricName, dbSettings.VersionStr, err)
+			lastSQLFetchError.Store(msg.MetricName+dbMetricJoinStr+fmt.Sprintf("%v", dbVersion), time.Now().Unix())
 		}
 		if strings.Contains(err.Error(), "too old") {
 			return nil, nil
@@ -662,7 +662,7 @@ func FetchMetrics(ctx context.Context,
 		}
 	}
 
-	sql = mvp.GetSQL(dbpgVersion)
+	sql = mvp.GetSQL(dbVersion)
 
 	if sql == "" && !(msg.MetricName == specialMetricChangeEvents || msg.MetricName == recoMetricName) {
 		// let's ignore dummy SQL-s
@@ -670,19 +670,19 @@ func FetchMetrics(ctx context.Context,
 		return nil, nil
 	}
 
-	if (mvp.PrimaryOnly() && vme.IsInRecovery) || (mvp.StandbyOnly() && !vme.IsInRecovery) {
-		log.GetLogger(ctx).Debugf("[%s:%s] Skipping fetching of  as server not in wanted state (IsInRecovery=%v)", msg.DBUniqueName, msg.MetricName, vme.IsInRecovery)
+	if (mvp.PrimaryOnly() && dbSettings.IsInRecovery) || (mvp.StandbyOnly() && !dbSettings.IsInRecovery) {
+		log.GetLogger(ctx).Debugf("[%s:%s] Skipping fetching of  as server not in wanted state (IsInRecovery=%v)", msg.DBUniqueName, msg.MetricName, dbSettings.IsInRecovery)
 		return nil, nil
 	}
 
 	if msg.MetricName == specialMetricChangeEvents && context != contextPrometheusScrape { // special handling, multiple queries + stateful
-		CheckForPGObjectChangesAndStore(ctx, msg.DBUniqueName, vme, storageCh, hostState) // TODO no hostState for Prometheus currently
+		CheckForPGObjectChangesAndStore(ctx, msg.DBUniqueName, dbSettings, storageCh, hostState) // TODO no hostState for Prometheus currently
 	} else if msg.MetricName == recoMetricName && context != contextPrometheusScrape {
-		if data, err = GetRecommendations(ctx, msg.DBUniqueName, vme); err != nil {
+		if data, err = GetRecommendations(ctx, msg.DBUniqueName, dbSettings); err != nil {
 			return nil, err
 		}
 	} else if msg.Source == sources.SourcePgPool {
-		if data, err = FetchMetricsPgpool(ctx, msg, vme, mvp); err != nil {
+		if data, err = FetchMetricsPgpool(ctx, msg, dbSettings, mvp); err != nil {
 			return nil, err
 		}
 	} else {
@@ -691,9 +691,9 @@ func FetchMetrics(ctx context.Context,
 		if err != nil {
 			// let's soften errors to "info" from functions that expect the server to be a primary to reduce noise
 			if strings.Contains(err.Error(), "recovery is in progress") {
-				dbPgVersionMapLock.RLock()
-				ver := dbPgVersionMap[msg.DBUniqueName]
-				dbPgVersionMapLock.RUnlock()
+				MonitoredDatabasesSettingsLock.RLock()
+				ver := MonitoredDatabasesSettings[msg.DBUniqueName]
+				MonitoredDatabasesSettingsLock.RUnlock()
 				if ver.IsInRecovery {
 					log.GetLogger(ctx).Debugf("[%s:%s] failed to fetch metrics: %s", msg.DBUniqueName, msg.MetricName, err)
 					return nil, err
@@ -718,7 +718,7 @@ func FetchMetrics(ctx context.Context,
 
 		log.GetLogger(ctx).WithFields(map[string]any{"source": msg.DBUniqueName, "metric": msg.MetricName, "rows": len(data)}).Info("measurements fetched")
 		if regexIsPgbouncerMetrics.MatchString(msg.MetricName) { // clean unwanted pgbouncer pool stats here as not possible in SQL
-			data = FilterPgbouncerData(ctx, data, md.GetDatabaseName(), vme)
+			data = FilterPgbouncerData(ctx, data, md.GetDatabaseName(), dbSettings)
 		}
 
 		ClearDBUnreachableStateIfAny(msg.DBUniqueName)
@@ -732,9 +732,9 @@ func FetchMetrics(ctx context.Context,
 send_to_storageChannel:
 
 	if (opts.Measurements.RealDbnameField > "" || opts.Measurements.SystemIdentifierField > "") && msg.Source == sources.SourcePostgres {
-		dbPgVersionMapLock.RLock()
-		ver := dbPgVersionMap[msg.DBUniqueName]
-		dbPgVersionMapLock.RUnlock()
+		MonitoredDatabasesSettingsLock.RLock()
+		ver := MonitoredDatabasesSettings[msg.DBUniqueName]
+		MonitoredDatabasesSettingsLock.RUnlock()
 		data = AddDbnameSysinfoIfNotExistsToQueryResultData(data, ver, opts)
 	}
 
@@ -745,16 +745,16 @@ send_to_storageChannel:
 	if fromCache {
 		log.GetLogger(ctx).Infof("[%s:%s] loaded %d rows from the instance cache", msg.DBUniqueName, msg.MetricName, len(cachedData))
 		return []metrics.MeasurementMessage{{DBName: msg.DBUniqueName, MetricName: msg.MetricName, Data: cachedData, CustomTags: md.CustomTags,
-			MetricDef: mvp, RealDbname: vme.RealDbname, SystemIdentifier: vme.SystemIdentifier}}, nil
+			MetricDef: mvp, RealDbname: dbSettings.RealDbname, SystemIdentifier: dbSettings.SystemIdentifier}}, nil
 	}
 	return []metrics.MeasurementMessage{{DBName: msg.DBUniqueName, MetricName: msg.MetricName, Data: data, CustomTags: md.CustomTags,
-		MetricDef: mvp, RealDbname: vme.RealDbname, SystemIdentifier: vme.SystemIdentifier}}, nil
+		MetricDef: mvp, RealDbname: dbSettings.RealDbname, SystemIdentifier: dbSettings.SystemIdentifier}}, nil
 
 }
 
 var pgBouncerNumericCountersStartVersion = 01_12_00 // pgBouncer changed internal counters data type in v1.12
 
-func FilterPgbouncerData(ctx context.Context, data metrics.Measurements, databaseToKeep string, vme DBVersionMapEntry) metrics.Measurements {
+func FilterPgbouncerData(ctx context.Context, data metrics.Measurements, databaseToKeep string, vme MonitoredDatabaseSettings) metrics.Measurements {
 	filteredData := make(metrics.Measurements, 0)
 
 	for _, dr := range data {

--- a/src/reaper/recommendations.go
+++ b/src/reaper/recommendations.go
@@ -27,7 +27,7 @@ const (
 var specialMetrics = map[string]bool{recoMetricName: true, specialMetricChangeEvents: true, specialMetricServerLogEventCounts: true}
 var regexIsPgbouncerMetrics = regexp.MustCompile(specialMetricPgbouncer)
 
-func GetAllRecoMetricsForVersion(vme DBVersionMapEntry) (map[string]metrics.Metric, error) {
+func GetAllRecoMetricsForVersion(vme MonitoredDatabaseSettings) (map[string]metrics.Metric, error) {
 	mvpMap := make(map[string]metrics.Metric)
 	metricDefMapLock.RLock()
 	defer metricDefMapLock.RUnlock()
@@ -43,7 +43,7 @@ func GetAllRecoMetricsForVersion(vme DBVersionMapEntry) (map[string]metrics.Metr
 	return mvpMap, nil
 }
 
-func GetRecommendations(ctx context.Context, dbUnique string, vme DBVersionMapEntry) (metrics.Measurements, error) {
+func GetRecommendations(ctx context.Context, dbUnique string, vme MonitoredDatabaseSettings) (metrics.Measurements, error) {
 	retData := make(metrics.Measurements, 0)
 	startTimeEpochNs := time.Now().UnixNano()
 

--- a/src/reaper/types.go
+++ b/src/reaper/types.go
@@ -25,7 +25,7 @@ const (
 
 )
 
-type MetricFetchMessage struct {
+type MetricFetchConfig struct {
 	DBUniqueName        string
 	DBUniqueNameOrig    string
 	MetricName          string
@@ -41,7 +41,7 @@ type ChangeDetectionResults struct { // for passing around DDL/index/config chan
 	Dropped int
 }
 
-type DBVersionMapEntry struct {
+type MonitoredDatabaseSettings struct {
 	LastCheckedOn    time.Time
 	IsInRecovery     bool
 	VersionStr       string

